### PR TITLE
Update: generate Mac bundle on install rather than packaging

### PR DESCRIPTION
**Pull request by danchr**
This is a major overhaul of how OpenTTD is packaged for macOS. Rather than using a disk image, it's now a Zip archive. The primary motiviation for this was:

1) Disk images are annoying.
2) Technically, creating the bundle _is_ part of the build.

I also added an option for disabling the app bundle “fixup”. This should be useful for developers or packagers such as MacPorts — which was how I started looking at this.

### CODE REVIEW

**Comment by andythenorth** [comment]
Is .dmg not the macOS convention any more?

**Comment by danchr** [comment]
> Is .dmg not the macOS convention any more?

I can't speak to this authoritatively, but I always found them a bit annoying. Since this is actually getting into UX, I tried to find a some references for my hunch:

* John Gruber complained about this in 2009: [How Should Mac Apps Be Distributed?](https://daringfireball.net/2009/09/how_should_mac_apps_be_distributed)
* Apple _did_ recommend images at some point, I believe, but the current guide only speaks of archives: [Distribute outside the Mac App Store (macOS)](https://help.apple.com/xcode/mac/current/#/dev033e997ca).
* The only relevant post I could find on Stack Overflow isn't terribly informative — and from 2010: [DMG or Zip file for distribution to Macs?](https://stackoverflow.com/questions/3954506/dmg-or-zip-file-for-distribution-to-macs)

Apple's guide only mentions archives and installer packages. Safari transparently extracts archives, and the Installer offers to move packages to the trash after processing, so either leads to a more pleasant experience in my opinion.

Unfortunately, all that won't matter much unless the product is signed somehow, as that triggers a dialog that prevents launching the application unless you right-click it.

One other advantage is that creating an archive is faster than creating the disk image.

**Comment by andythenorth** [comment]
I'm not aware of any advantage of .dmg other than convention.  Historically they were used to preserve the resource forks, folder layout etc.  I don't know if that still holds.

I found the same links you did, and nothing current from Apple recommending .dmg.

File: src/os/unix/unix.cpp [hunk 1]
```diff
@@ -250,6 +251,9 @@ int CDECL main(int argc, char *argv[])
 	if (argc >= 2 && strncmp(argv[1], "-psn", 4) == 0) {
 		argv[1] = nullptr;
 		argc = 1;
+
+		/* change the directory to the application bundle */
+		chdir(dirname(argv[0]));
```
Comment (LordAro) [inline src/os/unix/unix.cpp]: This might break some assumptions about configfile location

Comment (danchr) [reply src/os/unix/unix.cpp]: Yeah, re-reading `launch.sh`, I can see I got it wrong:

```shell
#!/bin/sh		

ROOT_DIR=$(dirname "$0")/..		

export DYLD_LIBRARY_PATH=${ROOT_DIR}/Frameworks		

cd ${ROOT_DIR}/Resources		
exec ./openttd "$@"
```

But the fact that OpenTTD tries to write to its own application bundle is rather surprising, and not guaranteed to work in all situations. But addressing that may be too involved; I'll have to do something else…

(The `DYLD_LIBRARY_PATH` line isn't required in the CMake build, but the rest probably is.)

**Comment by glx22** [comment]
I see you mentioned "fixup", it looks like [nightly build](https://dev.azure.com/openttd/OpenTTD/_build/results?buildId=5998&view=logs&j=a5e52b91-c83f-5429-4a68-c246fc63a4f7&t=2dac33d1-b0ca-5feb-3c36-18b56319df50&l=47) fails at this step.
Maybe you can help me to understand what's wrong.

File: src/os/unix/unix.cpp (hunk 1 above)
Comment (danchr) [reply src/os/unix/unix.cpp]: Turns out this change not only isn't needed, but actually [doesn't even work](https://github.com/macports/macports-ports/blob/c2f29c31270d13afba23ef8eadb506e8770a10ec/_resources/port1.0/group/app-1.0.tcl#L222) on 10.9 and later. So I never actually tested it by launching…

Still, though, OpenTTD appears to detect its location just fine when launched as-is.

**Comment by danchr** [comment]
> I see you mentioned "fixup", it looks like [nightly build](https://dev.azure.com/openttd/OpenTTD/_build/results?buildId=5998&view=logs&j=a5e52b91-c83f-5429-4a68-c246fc63a4f7&t=2dac33d1-b0ca-5feb-3c36-18b56319df50&l=47) fails at this step.
> Maybe you can help me to understand what's wrong.

Well, it does appear to run into some permissions issues _after_ copying dependencies — which seems a bit odd. What are the user accounts used during the build? Does the `install` happen with the same user as the actual build?

**Comment by glx22** [comment]
I fixed the issue with https://github.com/OpenTTD/OpenTTD/commit/b145ee310c042212ce5d6520383f7ef2e5691c21 but `make install` is still broken in master branch, and I think this PR will solve that.

**Comment by danchr** [comment]
> I fixed the issue with [b145ee3](https://github.com/OpenTTD/OpenTTD/commit/b145ee310c042212ce5d6520383f7ef2e5691c21) but `make install` is still broken in master branch, and I think this PR will solve that.

I pushed a change so that it should try this out on my branch. We'll see if it addresses this. One of the motivations for this change was that CPack seemed a bit crufty and weird compared to generating the bundle during build.

**Comment by danchr** [comment]
> I fixed the issue with [b145ee3](https://github.com/OpenTTD/OpenTTD/commit/b145ee310c042212ce5d6520383f7ef2e5691c21) but `make install` is still broken in master branch, and I think this PR will solve that.

Nope; the change I pushed yesterday gets the exact same error, sadly. But I've pushed another applying the same fix you did.

**Comment by danchr** [comment]
I've rebased the change onto `master`; could you please take another look?

**Comment by TrueBrain** [comment]
Sorry it has been a while; we are currently working on migrating our release workflow from Azure Pipelines to GitHub Actions. During so, we also prepared the Mac builds so they can produce ARM builds and fixed several Mac OS related bundle issues.

During this work, we found out that depending on CPack saves us a lot of time and effort. And although I appreciate your stance on not-liking-`dmg`-files, I am not convinced that adding the additional code (which we would have to maintain) is worth it. The reason we switched to CMake to start with was to avoid having to maintain all the quirks OSes have ourselves :)

In result, I am leaning towards closing this PR in favour of CPack `dmg`. This is mainly because the arguments you bring forth `dmg` is not a nice way to distribute are rather thin, and when we look around (as a non-Mac-user), we still see many projects distribute their applications in `dmg` format.
This is not a: we-are-never-going-to-do-this, but currently I do not see in the balance of "more custom code" vs "do what-ever CPack does" why the former should win :) I can be convinced otherwise ;)

**Pull request closed without merging**